nova_vm/src/ecmascript/builtins/fundamental_objects/function_objects/function_constructor.rs

```diff
@@ -52,3 +74,172 @@ impl FunctionConstructor {
+    // avoid code injection, but oxc doesn't have a public API to do that.
+    // Instead, we parse the source string as a script, and throw unless it has
+    // exactly one statement which is a function declaration of the right kind.
+    let (function, source_code) = {
+        let mut function = None;
+        let mut source_code = None;
+
+        // SAFETY: The safety requirements are that the SourceCode cannot be
+        // GC'd before the program is dropped. If this function returns
+        // successfully, then the program's AST and the SourceCode will both be
+        // kept alive in the returned function object.
+        let parsed_result =
+            unsafe { SourceCode::parse_source(agent, source_string, Default::default()) };
+
+        if let Ok((program, sc)) = parsed_result {
+            source_code = Some(sc);
+            if program.hashbang.is_none()
+                && program.directives.is_empty()
+                && program.body.len() == 1
+            {
+                if let oxc_ast::ast::Statement::FunctionDeclaration(funct) = &program.body[0] {
+                    if kind.function_matches_kind(funct) {
+                        // SAFETY: the Function is inside a oxc_allocator::Box, which will remain
+                        // alive as long as `source_code` is kept alive. Similarly, the inner
+                        // lifetime of Function is also kept alive by `source_code`.`
+                        function = Some(unsafe {
+                            std::mem::transmute::<
+                                &oxc_ast::ast::Function,
+                                &'static oxc_ast::ast::Function,
+                            >(funct)
+                        });
+                    }
+                }
+            }
+        }
+
+        if let Some(function) = function {
+            (function, source_code.unwrap())
+        } else {
+            return Err(agent.throw_exception_with_static_message(
+                ExceptionType::SyntaxError,
+                "Invalid function source text.",
+            ));
+        }
+    };
+
+    let params = OrdinaryFunctionCreateParams {
+        function_prototype: get_prototype_from_constructor(
+            agent,
+            constructor,
+            kind.intrinsic_prototype(),
+        )?,
+        source_code: Some(source_code),
+        source_text: function.span,
+        parameters_list: &function.params,
+        body: function.body.as_ref().unwrap(),
+        is_concise_arrow_function: false,
+        is_async: function.r#async,
+        is_generator: function.generator,
+        lexical_this: false,
+        env: EnvironmentIndex::Global(agent.current_realm().global_env.unwrap()),
+        private_env: None,
+    };
+    let f = ordinary_function_create(agent, params);
+
+    set_function_name(
+        agent,
+        f,
+        BUILTIN_STRING_MEMORY.anonymous.to_property_key(),
+        None,
+    );
+    match kind {
```

Comment:
**suggestion:** Move these branches to the callers, as they statically know what their kind is.

Reply:
Done.